Transaction list tests

piotr-iohk opened the pull request:
# Issue Number

#466 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] API tests
- [x] CLI tests


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

Commit:
review alignment in cli tests

Commit:
input, output, insertedAt lenses in DSL

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1019,26 +1028,71 @@ spec = do
+        -- let outs = getFromResponseList @(ApiTransaction t) 0 outputs r
+        -- let ins = getFromResponseList 0 inputs r
+        -- length outs `shouldBe` 10
+        -- length ins `shouldBe` 10
```
... I'd like to assert for `inputs` and `outputs` size... But it doesn't type-check. I think the `input` and `output` lenses are wrong (getFromResponseList works fine on other lenses)

jonathanknowles commented on lib/core-integration/src/Test/Integration/Framework/DSL.hs:
```diff
@@ -643,6 +658,39 @@ direction =
+    _set :: HasType (Maybe ApiBlockData) s => (s, (Maybe Text)) -> s
+    _set (s, v) = case v of
+        Just t -> Just (set typed (ApiBlockData (textToUTCTime t) mempty) s)
+        Nothing -> Nothing
```
Hi @piotr-iohk!

Given that your other lenses have non-textual types, I wonder why we couldn't instead define `insertedAt` with the following type:

```hs
insertedAt :: HasType (Maybe ApiBlockData) s => Lens' s (Maybe UTCTime)
```

That way, we wouldn't need to have the silent default in the `textToUTCTime` function. The reason I worry about silent defaults is that they could mask errors introduced by people modifying the code in future.

However, if using the `Text` type is more useful, then perhaps we could have `_set` use a default of `Nothing` if the given text is invalid, which is compatible with the original type you've used (`Maybe Text`).

What do you think?

piotr-iohk replied:
I was going initially for `Text` as I was thinking it'd be easier to assert in tests, but maybe `UTCTime` will be better, we'll see when I get to the tests part... First I need to make this freaking lens work ;)

> Given that your other lenses have non-textual types,
 
btw. actually there is already a date-time lens that is a `Text` type -> `passphraseLastUpdate`

>a default of Nothing if the given text is invalid,

If I go for `Text` I'll try this.

piotr-iohk replied:
OK, went for:

```haskell
insertedAt :: HasType (Maybe ApiBlockData) s => Lens' s (Maybe UTCTime)
```

Commit:
Tx list tests in API

Commit:
rebase after #613

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1043,338 @@ spec = do
+            ]
+
+    -- This scenario covers the following matrix of cases. Cases where generated
+    -- using one of parwise test cases generation tools available online.
```
Spotted a couple of typos:
> "Cases **were** generated using one of the **pairwise**..."

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1043,338 @@ spec = do
+                    <> (toParam t1)
+                    <> "&end="
+                    <> (toParam t2)
+                    <> "&order=ascending"
```
Perhaps the `toQueryString` function would be helpful here (and in other places below)?

https://github.com/input-output-hk/cardano-wallet/pull/613/files#diff-213a7e10be74fd4570beffc3464fd728R825

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1043,338 @@ spec = do
+                , ( "?start="
+                    <> (toParam t1)
+                    <> "&end="
+                    <> (toParam $ addUTCTime 1 t2)
```
If you're trying to get the next `UTCTime` value after `t2` here, then you could use the following construction:
```hs
<> (toParam $ addUTCTime (succ 0) t2)
```
In this case, the `Enum` instance guarantees not to skip any intermediate values.

On the other hand, specifying `1` here will actually get you an increment of `10^12` picoseconds, because while `NominalDiffTime` does have a `Num` instance, it multiplies the value internally using the functionality provided by `Data.Fixed`:

http://hackage.haskell.org/package/base-4.12.0.0/docs/src/Data.Fixed.html#line-100

piotr-iohk replied:
> If you're trying to get the next UTCTime value after t2 here, then you could use the following construction:

Actually I was going for getting the next UTCTime incremented/decremented by exactly 1 second (i.e. `10^12` picoseconds), since I think it'd be the most common usecase. 

We already have `TIME_RANGE_LIMIT` tests which tests small ranges (in `picoseconds`) implemented by you, so I we cover that. I think that both are somewhat complementary.

jonathanknowles replied:
Fair enough! My question was mostly to figure out your intentions here. I think the existing tests provide coverage that's good enough for our needs. :)

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1043,338 @@ spec = do
+                , ( "?start="
+                    <> (toParam t1)
+                    <> "&end="
+                    <> (toParam $ addUTCTime (-1) t2)
```
Similarly to the previous comment, assuming we want the closest time to `t2` that is smaller than `t2`, we have to use `pred 0` rather than `-1`:
```hs
<> (toParam $ addUTCTime (pred 0) t2)
```

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1043,338 @@ spec = do
+                    )
+                -- 5
+                , ( "?start="
+                    <> (toParam $ addUTCTime 1 t1)
```
Scrolling down, it looks like there are a lot of these cases! I had a couple of ideas I wanted to share with you:

1. We could use named constants:

    ```hs
    onePicosecond :: NominalDiffTime
    onePicosecond = succ 0
    ```

Or

2. We could define a pair of functions to find the previous or next `UTCTime`:

    ```hs
    -- | For a given time 't0', get the closest representable time 't1' to 't0'
    --   for which 't0 < t1'.
    utcTimeSucc :: UTCTime -> UTCTime
    utcTimeSucc = addUTCTime (succ 0)

   -- | For a given time 't0', get the closest representable time 't1' to 't0'
    --   for which 't1 < t0'.
    utcTimePred :: UTCTime -> UTCTime
    utcTimePred = addUTCTime (pred 0)

    ```

What do you think?

I suspect using the functions might be clearer. (It's unfortunate that `UTCTime` cannot provide an `Enum` instance.)

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1043,338 @@ spec = do
+                      ]
+                    )
+                  ,
+                    ( "?start=2012-09-25T10:15:00Z&end=2016-11-21"
```
I suspect that the `:` character should be URL-encoded here.

Perhaps a good use case for the `toQueryString` function? (This handles URL encoding automatically.)

https://github.com/input-output-hk/cardano-wallet/pull/613/files#diff-213a7e10be74fd4570beffc3464fd728R825

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1043,338 @@ spec = do
         \ctx -> do
               w <- emptyWallet ctx
               let startTime = "2009-09-09T09:09:09Z"
               let endTime = "2001-01-01T01:01:01Z"
```
Perhaps it would also be worth adding a case that tests the inequality a bit more thoroughly.

After all, we do actually allow `start == end`.

piotr-iohk replied:
> After all, we do actually allow start == end.

This case is missing indeed.

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1233,3 +1564,42 @@ spec = do
+        r <- request @([ApiTransaction t]) ctx (listTxEp wDest "?order=ascending")
+            Default Empty
+        let indexes = [0..(length txAmounts - 1)]
+        let txTimes = map (\x -> fromJust $ getFromResponseList x insertedAtTime r) indexes
```
Some lines in this PR are a bit longer than our coding standard allows.

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

Perhaps have a quick scan over the files changed in this PR?

For the above case, perhaps we can format like this:
```hs
        let txTimes = map
                (\x -> fromJust $ getFromResponseList x insertedAtTime r)
                indexes
```

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/CLI/Transactions.hs:
```diff
@@ -648,76 +648,76 @@ spec = do
-                        <> " in "
-                        <> show mOrder
-                        <> " order "
+                      <> "listing transactions from "
```
The previous :arrow_right: indentation of 4 spaces was actually correct, according to our standard. (Perhaps this is a spurious diff?)

piotr-iohk replied:
I think so, the alignment _fix_ here was mainly about moving `TRANS_LIST` tests out of the `describe "TRANS_ESTIMATE_09 - Invalid amount"` _suite_.

jonathanknowles reviewed (commented):
Looks good so far.

My main comment would be that we should probably be careful with how we increment and decrement `UTCTime` values.

Rather than using `addUTCTime 1` and `addUTCTime (-1)` (both of which skip `10^12` picoseconds), we should probably be using `addUTCTime (succ 0)` (which finds the next consecutive value) and `addUTCTime (pred 0)` (which finds the previous consecutive value) respectively.

We could probably define helper functions for this common pair of use cases.

Commit:
use toQueryString, fix indentation, add start == end case

Commit:
fixtureWalletManyTxs, expectCliListItemFieldBetween in DSL

Commit:
Tx list tests in CLI

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1040,373 @@ spec = do
+    -- +----+----------+----------+------------+--------------+
+    --     |  start   |   end    |   order    |    result    |
+    -- +----+----------+----------+------------+--------------+
+    --   1 | edge     | edge     | ascending  | 2 ascending  |
```
Seemed a bit weird to me at first, but it's also very reassuring to see all of these explicit cases!

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1040,373 @@ spec = do
+                        ]
+                  ,  [ expectListSizeEqual 2
+                     , expectListItemFieldEqual 0 amount a1
+                     , expectListItemFieldEqual 1 amount a2
```
A bit new to this, but would it not be cleaner to check for equality against `[a1, a2]` (somehow), instead of 1) check the size and 2) check element at each index?

piotr-iohk replied:
>cleaner to check for equality against [a1, a2] (somehow)

Not sure if I follow, o, maybe, I have no other idea how to make it simpler/more readable :)

Yes, it simply checks if there are two and only two txs on the returned list and asserts they are returned in expected order (i.e. amount a1 on index 0, and amount a2 on index 1)

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -560,7 +575,7 @@ spec = do
         expectErrorMessage (errMsg404NoWallet $ w ^. walletId) r
 
-    describe "TRANS_CREATE_08 - v2/wallets/{id}/transactions - Methods Not Allowed" $ do
+    describe "TRANS_CREATE_08, TRANS_LIST_04 - v2/wallets/{id}/transactions - Methods Not Allowed" $ do
```
This line is quite a bit longer than the limit in our coding standards:

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

There's a strategy for fixing this here:

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#strategy-3-break-up-long-string-literals

Suggestion: insert a break between the title and the description, like so:
```suggestion
    describe
        "TRANS_CREATE_08, TRANS_LIST_04 - \
        \v2/wallets/{id}/transactions - Methods Not Allowed" $ do
```

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1040,373 @@ spec = do
+    -- using one of pairwise test cases generation tools available online.
+    -- +----+----------+----------+------------+--------------+
+    --     |  start   |   end    |   order    |    result    |
+    -- +----+----------+----------+------------+--------------+
```
Somehow the alignment ended up broken here. :)

```suggestion
    -- +---+----------+----------+------------+--------------+
    --     |  start   |   end    |   order    |    result    |
    -- +---+----------+----------+------------+--------------+
```

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1222,3 +1585,7 @@ spec = do
+
+    plusOneSecond, minusOneSecond :: UTCTime -> UTCTime
+    plusOneSecond = addUTCTime 1
+    minusOneSecond = addUTCTime (-1)
```
I like the extra clarity that these definitions provide.

KtorZ commented on lib/core-integration/src/Test/Integration/Framework/DSL.hs:
```diff
@@ -763,6 +840,55 @@ fixtureWalletWith ctx coins0 = do
+  -- ^ List of tx amounts
+  -> IO (ApiWallet, ApiWallet, [(UTCTime, Natural)])
+  -- ^ (wSrc, wDest, wDestTxList)
+fixtureWalletManyTxs ctx txAmounts = do
```
This seems redundant with:

```hs
fixtureWalletWith
    :: forall t. (EncodeAddress t, DecodeAddress t)
    => Context t
    -> [Natural]
    -> IO ApiWallet
```

except for the return types that seems rather awkward to me. I'd rather see two calls to create the transactions and then list them than doing both steps in a single _rather cryptic_ function :grimacing:

piotr-iohk replied:
Actually `fixtureWalletWith` provides wallet with certain UTxO. It construct tx with batches of 10 inputs each so `fixtureWalletWith ctx [1,2]` is a wallet with one tx, but 2 UTxos.

KtorZ replied:
Indeed, still doable to do: 

```hs
fixtureWalletWith ctx $ mconcat
    [ replicate 10 1
    , replicate 10 2
    ]
```

:grimacing: ? What do you think :sweat_smile: ?

piotr-iohk replied:
ok, removed`fixtureWalletManyTxs`

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1040,373 @@ spec = do
+        let (feeMin, feeMax) = ctx ^. feeEstimator $ TxDescription
+                { nInputs = 1
+                , nOutputs = 1
+                }
```
I don't think asserting fees is really relevant for that unit test. If we're looking into behavior of the list transaction endpoint, we should reduce the noise around and only checks what's relevant to listing transactions.

piotr-iohk replied:
It was just to assert for the amount of Outgoing transaction, but OK, removing for the sake of making the test clearer.

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1040,373 @@ spec = do
+        -- Incoming tx inputs and outpus size
+        let outsIn = getFromResponseList 1 outputs r
+        let insIn = getFromResponseList 1 inputs r
+        length (outsIn :: [AddressAmount t]) `shouldBe` 1000
```
Whola. This sounds like something that is very specific to the way the http-bridge is set up. I wouldn't expect fixtures to yield such transactions! 
Generally speaking, we shouldn't make any assertions on the tx structure of the faucet wallets.

piotr-iohk replied:
removed

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1040,373 @@ spec = do
+                     , expectListItemFieldEqual 0 amount a1
+                     , expectListItemFieldEqual 1 amount a2
+                     ]
+                  )
```
For better readability, I'd recommend to switch from a tuple to a record type and used the constructor with the named fields to make it easy for the reader to go through the matrix :+1:
Tuples are the poor man's sum types, acceptable for short arguments. Here, since we have many cases with rather long arguments, a record would combine them in a much more readable way!

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1040,373 @@ spec = do
+    --         if (title == "40 chars hex") then
+    --             expectErrorMessage (errMsg404NoWallet $ T.pack walId) r
+    --         else
+    --             expectErrorMessage errMsg404NoEndpoint r
```
Commented-out code :thinking: ?

Commit:
code review alignments

Commit:
simplify and make fixtureWalletManyTxs less cryptic

Commit:
remove fixtureWalletManyTxs

Anviking committed:
Fix stylish-haskell

Commit:
add TestCase data type instead of tuple in test matrix

Commit:
fix TestCase

Commit:
remove InLedger assertion for TRANS_LIST_01

jonathanknowles committed:
Fix formatting in `Test.Integration.Scenario.API.Transactions`.

Indents should now be 4 spaces per indent.

(Previously, there was a bit of a mixture of different idents.)

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/CLI/Transactions.hs:
```diff
@@ -648,76 +658,278 @@ spec = do
+              length <$> [oJson1, oJson2] `shouldSatisfy` all (== 0)
+
+    it "TRANS_LIST_RANGE_03 - \
+       \Transaction at time t is NOT selected by range (..., t - 𝛿t)" $
```
```suggestion
       \Transaction at time t is NOT selected by range (..., t - 𝛿t]" $
```
since it corresponds to `t - 𝛿t <= x`. Same in the case above.

Anviking replied:
Suggested fix for both: cd93039418f64e9c7d839e10dcb076e6175d56df

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1040,409 @@ spec = do
+                        ]
+                    , assertions =
+                        [ expectListSizeEqual 2
+                        , expectListItemFieldEqual 0 amount a2
```
I was curious if it was possible to remove `expectListItemFieldEqual 0 amount` in favour for like `expectFieldEqual (0_ . amount)`, but I didn't figure out how with generic-lens. Arguably that would be less readable also / possibly against the original design-principles of the DSL.

Anviking reviewed (approved):
I think this looks good to me.

Anviking committed:
Denote intervals as closed instead of open in test descriptions

Anviking committed:
Update .nix

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -1025,26 +1040,409 @@ spec = do
+        let a2 = sum $ replicate 10 2
+        w <- fixtureWalletWith ctx $ mconcat
+                [ replicate 10 1
+                , replicate 10 2
```
Actually… what is this doing?

This is relying on [`We only have 10 UTxO available in each faucet`](https://github.com/input-output-hk/cardano-wallet/blob/37be502ca7443f53316b75caedf7623c16ce6f95/lib/core-integration/src/Test/Integration/Framework/DSL.hs#L794-L810) to make sure the wallet gets set up using two different transactions?

piotr-iohk replied:
That was exactly the intent (following [this](https://github.com/input-output-hk/cardano-wallet/pull/608#discussion_r313385024) suggestion).
